- Name: Isaac Ndirangu Muturi
- Email: ndirangumuturi749@gmail.com

## Week 5 Homework 

For this homework we will be using the FHV 2019-10 data found here. [FHV Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz)


### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?

> [!NOTE]
> To install PySpark follow this [guide](https://github.com/DataTalksClub/data-engineering-zoomcamp/blob/main/05-batch/setup/pyspark.md)

In [1]:
import pyspark
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


24/03/04 19:01:53 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [2]:
spark

In [4]:
print(pyspark.__version__)


3.3.2


### Question 2: 

**FHV October 2019**

Read the October 2019 FHV into a Spark Dataframe with a schema as we did in the lessons.

Repartition the Dataframe to 6 partitions and save it to parquet.

What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.

- 1MB
- 6MB
- 25MB
- 87MB

In [5]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz


--2024-03-04 19:04:48--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz
Resolving github.com (github.com)... 140.82.112.3
Connecting to github.com (github.com)|140.82.112.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-44d1-a138-4e8ea3c3a3b6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240304%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240304T190448Z&X-Amz-Expires=300&X-Amz-Signature=9775d700250efeb7cd38589d3edeea30d4c84ee5d91339fbf0805ed0adbd7eb9&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhv_tripdata_2019-10.csv.gz&response-content-type=application%2Foctet-stream [following]
--2024-03-04 19:04:48--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-

In [8]:
!gunzip -kf fhv_tripdata_2019-10.csv.gz


In [9]:
from pyspark.sql import types

# Define the schema for FHV data
fhv_schema = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropoff_datetime', types.TimestampType(), True),
    types.StructField('PULocationID', types.IntegerType(), True),
    types.StructField('DOLocationID', types.IntegerType(), True),
    types.StructField('SR_Flag', types.StringType(), True)
])

In [12]:
year = 2019
month = 10

input_path = f'../data/raw/fhv/{year}/{month:02d}/'
output_path = f'../data/pq/fhv/{year}/{month:02d}/'

# Read FHV data with specified schema
df_fhv = spark.read \
    .option("header", "true") \
    .schema(fhv_schema) \
    .csv(input_path)

# Repartition the DataFrame to 6 partitions and save it to parquet
df_fhv \
    .repartition(6) \
    .write.parquet(output_path)


24/03/04 19:20:11 WARN CSVHeaderChecker: Number of column in CSV header is not equal to number of fields in the schema:
 Header length: 7, schema size: 6
CSV file: file:///home/ndirangu749/Data_Engineering_Zoomcamp_2024/module_5_batch_processing/data/raw/fhv/2019/10/fhv_tripdata_2019-10.csv.gz


In [14]:
output_path

'../data/pq/fhv/2019/10/'

In [17]:
!ls -lh ../data/pq/fhv/2019/10/

total 36M
-rw-r--r-- 1 ndirangu749 ndirangu749    0 Mar  4 19:20 _SUCCESS
-rw-r--r-- 1 ndirangu749 ndirangu749 6.0M Mar  4 19:20 part-00000-bda4c532-3691-4eee-a6bc-49d32df84646-c000.snappy.parquet
-rw-r--r-- 1 ndirangu749 ndirangu749 6.0M Mar  4 19:20 part-00001-bda4c532-3691-4eee-a6bc-49d32df84646-c000.snappy.parquet
-rw-r--r-- 1 ndirangu749 ndirangu749 6.0M Mar  4 19:20 part-00002-bda4c532-3691-4eee-a6bc-49d32df84646-c000.snappy.parquet
-rw-r--r-- 1 ndirangu749 ndirangu749 6.0M Mar  4 19:20 part-00003-bda4c532-3691-4eee-a6bc-49d32df84646-c000.snappy.parquet
-rw-r--r-- 1 ndirangu749 ndirangu749 6.0M Mar  4 19:20 part-00004-bda4c532-3691-4eee-a6bc-49d32df84646-c000.snappy.parquet
-rw-r--r-- 1 ndirangu749 ndirangu749 6.0M Mar  4 19:20 part-00005-bda4c532-3691-4eee-a6bc-49d32df84646-c000.snappy.parquet


### Question 3: 

**Count records** 

How many taxi trips were there on the 15th of October?

Consider only trips that started on the 15th of October.

- 108,164
- 12,856
- 452,470
- 62,610

> [!IMPORTANT]
> Be aware of columns order when defining schema

In [18]:
df_fhv = spark.read.parquet('../data/pq/fhv/2019/10')

In [30]:
from pyspark.sql.functions import col

df_oct_15_trips = df_fhv.filter(col("pickup_datetime").cast("date") == "2019-10-15")
df_oct_15_trips.show(5)

+--------------------+-------------------+-------------------+------------+------------+-------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|
+--------------------+-------------------+-------------------+------------+------------+-------+
|              B00628|2019-10-15 08:20:19|2019-10-15 09:30:40|         238|         132|   null|
|              B03016|2019-10-15 15:47:56|2019-10-15 16:20:50|         264|          94|   null|
|              B02930|2019-10-15 13:39:43|2019-10-15 13:44:54|         264|         185|   null|
|              B03160|2019-10-15 17:30:00|2019-10-15 18:13:00|          51|          51|   null|
|              B01800|2019-10-15 13:44:00|2019-10-15 16:04:00|         264|         264|   null|
+--------------------+-------------------+-------------------+------------+------------+-------+
only showing top 5 rows



In [19]:
num_trips_oct_15 = df_oct_15_trips.count()

print("Number of taxi trips on the 15th of October:", num_trips_oct_15)

Number of taxi trips on the 15th of October: 62610


### Question 4: 

**Longest trip for each day** 

What is the length of the longest trip in the dataset in hours?

- 631,152.50 Hours
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours

In [ ]:
from pyspark.sql.functions import col, unix_timestamp

# Calculate the duration of each trip in seconds
df_fhv = df_fhv.withColumn("trip_duration_seconds", (unix_timestamp("dropoff_datetime") - unix_timestamp("pickup_datetime")))

In [ ]:
# Convert duration to hours
df_fhv = df_fhv.withColumn("trip_duration_hours", col("trip_duration_seconds") / 3600)

In [31]:
# To find the Length of the longest trip in the dataset (in hours):
# Find the maximum duration
max_duration = df_fhv.agg({"trip_duration_hours": "max"}).collect()
max_duration

[Row(max(trip_duration_hours)=631152.5)]

### Question 5: 

**User Interface**

Spark’s User Interface which shows the application's dashboard runs on which local port?

- 80
- 443
- 4040
- 8080

In [ ]:
# 4040

### Question 6: 

**Least frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)

Using the zone lookup data and the FHV October 2019 data, what is the name of the LEAST frequent pickup location Zone?</br>

- East Chelsea
- Jamaica Bay
- Union Sq
- Crown Heights North

In [26]:
# Load the zone lookup data into a DataFrame
zone_lookup_df = spark.read.option("header", "true").csv("taxi+_zone_lookup.csv")
zone_lookup_df.head(5)

[Row(LocationID='1', Borough='EWR', Zone='Newark Airport', service_zone='EWR'),
 Row(LocationID='2', Borough='Queens', Zone='Jamaica Bay', service_zone='Boro Zone'),
 Row(LocationID='3', Borough='Bronx', Zone='Allerton/Pelham Gardens', service_zone='Boro Zone'),
 Row(LocationID='4', Borough='Manhattan', Zone='Alphabet City', service_zone='Yellow Zone'),
 Row(LocationID='5', Borough='Staten Island', Zone='Arden Heights', service_zone='Boro Zone')]

In [ ]:
# Create a temporary view for the zone lookup data
zone_lookup_df.createOrReplaceTempView("zone_lookup")

In [28]:
joined_df = df_fhv.join(zone_lookup_df, df_fhv.PULocationID == zone_lookup_df.LocationID)
joined_df.show(5)

+--------------------+-------------------+-------------------+------------+------------+-------+---------------------+-------------------+----------+-------------+--------------------+------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|trip_duration_seconds|trip_duration_hours|LocationID|      Borough|                Zone|service_zone|
+--------------------+-------------------+-------------------+------------+------------+-------+---------------------+-------------------+----------+-------------+--------------------+------------+
|              B00706|2019-10-11 10:27:51|2019-10-11 10:45:37|          23|         221|   null|                 1066| 0.2961111111111111|        23|Staten Island|Bloomfield/Emerso...|   Boro Zone|
|              B00256|2019-10-07 04:55:42|2019-10-07 05:14:21|         264|         264|   null|                 1119|0.31083333333333335|       264|      Unknown|                  NV|         N/A|
|         

In [29]:
# Count occurrences of each zone
zone_counts = joined_df.groupBy("Zone").count()
zone_counts.show(5)

+--------------------+-----+
|                Zone|count|
+--------------------+-----+
|           Homecrest| 1295|
|              Corona| 7175|
|    Bensonhurst West| 1880|
|         Westerleigh| 1317|
|Charleston/Totten...| 2533|
+--------------------+-----+
only showing top 5 rows



In [23]:
# Find the least frequent pickup location zone
least_frequent_zone = zone_counts.orderBy(col("count")).first()
least_frequent_zone

Row(Zone='Jamaica Bay', count=1)